### DISCUSSION TITLE
Make a file using make.include to run a pure C code

### DISCUSSION THREAD

**Discussion by salmeghlef** [contiki-ng/contiki-ng#2239] [General]
Hello everyone, 
I have a CoAP_server.c code that was built using pure C and I am trying to convert it to .sky so it can fit for Cooja motes. Is it possible to run pure C code through make.include and compile it to .sky using >> make TARGET=sky or not? when I try to run, it says :pthread.h is not defined and so on so forth for all the libraries of pure C, please help

**Comment by salmeghlef** [comment]
The code files can be downloaded from the below links:

https://drive.google.com/file/d/1wqOXc2sExr2QFxqUBuHfQcGqJwoEtkhv/view?usp=sharing

https://drive.google.com/file/d/1KXByL1Yb2C5Gvt8OGBrB_yhsa02rbWpi/view?usp=sharing

**Comment by pjonsson** [comment]
The `pthread.h` file is for "POSIX threads", and that is not covered by the C standard (and Contiki-NG does not support pthreads).

On a more practical level, you would probably run out of code memory on sky even if you had a pthreads-implementation for Contiki-NG.

**Reply by salmeghlef** [reply]
Thanks PJonson for you kind reply, 
I have misunderstanding for "out of code memory on sky", if you can explain more to me or refer me to some resources even on youtube, that would be appreciated.

**Reply by pjonsson** [reply]
The firmware won't fit in the flash memory on the device.

It's a moot point though, there isn't a pthreads implementation for Contiki-NG.

**Reply by salmeghlef** [reply]
Got the point, thank you. Have you developed a Contiki based application before? what should I do with my codes above, do I have to modify it to fit for Contiki support libraries?

**Reply by pjonsson** [reply]
There is an example server in `examples/coap/coap-example-server`. You can probably look at that to figure out how you need to modify your CoAP_server.c.

**Comment by salmeghlef** [comment]
Thanks Peter,
Please refer me to a Contiki based application developer so I can cotnanct
him for further development rather than CoAP.
في جمعة، ٢٨ أكتوبر، ٢٠٢٢ في ١٢:٤٨ م، كتب Peter A. Jonsson <
***@***.***>:

**Comment by salmeghlef** [comment]
Dear Peter,
What is the differences between .native and .sky when building the code for the mote?

Is it possible to convert .native to .sky or not?

Thank you

**Reply by pjonsson** [reply]
They are different platforms:

https://docs.contiki-ng.org/en/develop/doc/platforms/native.html
https://docs.contiki-ng.org/en/develop/doc/platforms/sky.html

The answer to your question is that it depends, some code can be compiled for both targets, and some can't. If you're still wondering about pthreads, I don't think it's possible to get that running on Sky.

**Comment by salmeghlef** [comment]
How about the cooja motes Peter, is it possible to apply a .native file on cooja mote or not?

**Reply by pjonsson** [reply]
Native and Cooja are more similar than Native and Sky, but the answer is still that it depends.

If you are asking about pthreads, it might be possible to link the mote, but the Cooja simulation assumes the motes don't use pthreads.

**Comment by salmeghlef** [comment]
Peter, do you think using FIT-IoT lab platform may accept to install the C code including pthreads on one of listed devices?
![FIT-IoT devices](https://user-images.githubusercontent.com/45724031/199350095-c68a9f70-acbd-4636-9a37-11eebcc571f9.png)
